ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -15,17 +15,175 @@
+    -> ('a * ('b, exn) Result.result) list
+  (** [parallel ~__context lst] groups actions in [lst] by order number and
+   * executes all actions with same order number in parallel.
+   * In the current implementation binds and fmaps are executed after waiting for all tasks of same
+   * order number to complete.
+   * Returns the results from [lst], in an unspecified order, hence each item in [lst] is a tuple
+   * with arbitrary data (e.g. a VM ref) as first element, and the action as second:
+   * this way the results will be associated with some meaningful data.
+   *
+   * e.g. given : [task1_order0 >>= action1 , task2_order1, task3_order0 >>= action2]
+   * we first launch task1 and task3, when both task1 and task3 has finished we execute
+   * action1 and action2 recursively to completion.
+   * Only then we launch task2 and waits for its completion.
+  *)
+end = struct
+  (** structured like a Freer Monad *)
+  type +'a t =
+    | Completed of 'a
+        (** [Completed result] an action that has finished, sometimes refered to as 'pure' *)
+    | Task of API.ref_task * (task_result -> 'a t)
+        (** [Task(task, next)] a task, and a function to call when the task finishes (continuation).
+     * This is an impure action. *)
+
+  (* Constructors *)
+
+  let return x = Completed x
+
+  let ok x = Completed (Result.Ok x)
+
+  let fail e = Completed (Result.Error e)
+
+  let wrap f = try f () with e -> Backtrace.is_important e ; fail e
+
+  let task f = wrap (fun () -> Task (f (), return))
+
+  (* Operations *)
+
+  let rec fmap f = function
+    | Completed x ->
+        (* just map the result through [f] if we already have it *)
+        Completed (f x)
+    | Task (task, task_next) ->
+        (* If everything was immediately available then the result would be:
+       * [f (task_next (result_of_task task))]
+       * However we first need to wait for the task to finish,
+       * and when we receive the result from the task, we need to map it through
+       * [task_next] and then through [f].
+       * The type system helps to ensure that the pipeline below is correct *)
+        Task (task, fun x -> x |> task_next |> fmap f)
+
+
+  module Infix = struct
+    let rec ( >>= ) m k =
+      match m with
+      | Completed x -> k x
+      | Task (task, next) ->
+          (* similar reasoning as above, when we get the result we need to chain the computations,
+         * refer to http://okmij.org/ftp/Computation/free-monad.html for a deeper theoretical explanation *)
+          Task (task, fun x -> next x >>= k)
+  end
+
+  let eval ~__context = function
+    | Completed _ as t -> t
+    | Task (task, next) as t ->
+      match result_of_task ~__context task with
+      | None -> t
+      | Some result ->
+          log_and_ignore_exn (fun () -> Xapi_task.destroy ~__context ~self:task) ;
+          next result
+
+
+  let classify completed lst =
+    List.fold_left
+      (fun (completed, pending, tasks) -> function
```

Comment:
Could this be written without nesting `fun` and `function`?
```
(function
| .. ->
| .. ->
)
```

Reply:
yes, but I think you then have to write out (completed, pending, tasks) multiple times.